lib/xilinx/bitstream_writer.cc

```diff
@@ -0,0 +1,353 @@
+#include <fstream>
+#include <iostream>
+
+#include <absl/strings/str_cat.h>
+#include <absl/strings/str_split.h>
+#include <absl/time/clock.h>
+#include <absl/time/time.h>
+
+#include <prjxray/bit_ops.h>
+#include <prjxray/xilinx/architectures.h>
+#include <prjxray/xilinx/bitstream_writer.h>
+
+namespace prjxray {
+namespace xilinx {
+
+// Per UG470 pg 80: Bus Width Auto Detection
+template <>
+typename BitstreamWriter<Series7>::header_t BitstreamWriter<Series7>::header_{
+    0xFFFFFFFF, 0x000000BB, 0x11220044, 0xFFFFFFFF, 0xFFFFFFFF, 0xAA995566};
+
+template <typename ArchType>
+int BitstreamWriter<ArchType>::writeBitstream(
+    const typename ArchType::ConfigurationPackage& packets,
+    const std::string& part_name,
+    const std::string& frames_file,
+    const std::string& generator_name,
+    const std::string& output_file) {
+	std::ofstream out_file(output_file, std::ofstream::binary);
+	if (!out_file) {
+		std::cerr << "Unable to open file for writting: " << output_file
+		          << std::endl;
+		return 1;
+	}
+
+	BitstreamHeader bit_header(
+	    create_header(part_name, frames_file, generator_name));
+	out_file.write(reinterpret_cast<const char*>(bit_header.data()),
+	               bit_header.size());
+
+	auto end_of_header_pos = out_file.tellp();
+	auto header_data_length_pos =
+	    end_of_header_pos - static_cast<std::ofstream::off_type>(4);
+
+	BitstreamWriter<ArchType> out_bitstream_writer(packets);
+	for (uint32_t word : out_bitstream_writer) {
+		if (sizeof(typename ArchType::WordType) == 4) {
+			out_file.put((word >> 24) & 0xFF);
+			out_file.put((word >> 16) & 0xFF);
+		}
```

Comment:
Maybe restructure this as :
if (sizeof(typename ArchType::WordType) == 4) {
 <loop and write 32-bit words>
} else {
 <assert size(wordtype) == 2>
 <loop and write 16-bit words>
 <assert that upper 16-bit of 32 bit word is zero?>
}

Reply:
what about:
<pre>
       int bytes_per_word = sizeof(typename ArchType::WordType);
        for (uint32_t word : out_bitstream_writer) {
                for (int byte = bytes_per_word - 1; byte >= 0; byte--) {
                        out_file.put((word >> (byte * 8)) & 0xFF);
                }
        }
</pre>